app/routers/wallet.py

```diff
@@ -65,77 +147,30 @@ async def create_public_did():
             verkey=did_object["verkey"],
             paymentaddr="",
         ).dict()
-        r = requests.post(url, data=json.dumps(payload), headers={})
-        if r.status_code != 200:
-            error_json = r.json()
-            raise HTTPException(
-                status_code=r.status_code,
-                detail=f"Something went wrong.\nCould not write to StagingNet.\n{error_json}",
-            )
-        taa_response = await aries_agent_controller.ledger.get_taa()
-        logger.info(f"taa_response:\n{taa_response}")
-        if not taa_response["result"]:
-            error_json = taa_response.json()
-            raise HTTPException(
-                status_code=404,
-                detail=f"Something went wrong. Could not get TAA. {error_json}",
-            )
-        TAA = taa_response["result"]["taa_record"]
-        TAA["mechanism"] = "service_agreement"
-        accept_taa_response = await aries_agent_controller.ledger.accept_taa(TAA)
-        logger.info(f"accept_taa_response: {accept_taa_response}")
-        if accept_taa_response != {}:
-            error_json = accept_taa_response.json()
-            raise HTTPException(
-                status_code=404,
-                detail=f"Something went wrong. Could not accept TAA. {error_json}",
-            )
-        assign_pub_did_response = await aries_agent_controller.wallet.assign_public_did(
-            did_object["did"]
-        )
-        logger.info(f"assign_pub_did_response:\n{assign_pub_did_response}")
-        if (
-            not assign_pub_did_response["result"]
-            or assign_pub_did_response["result"] == {}
-        ):
-            error_json = assign_pub_did_response.json()
-            raise HTTPException(
-                status_code=500,
-                detail=f"Something went wrong.\nCould not assign DID. {error_json}",
-            )
-        get_pub_did_response = await aries_agent_controller.wallet.get_public_did()
-        logger.info(f"get_pub_did_response:\n{get_pub_did_response}")
-        if not get_pub_did_response["result"] or get_pub_did_response["result"] == {}:
-            error_json = get_pub_did_response.json()
-            raise HTTPException(
-                status_code=404,
-                detail=f"Something went wrong. Could not obtain public DID. {error_json}",
-            )
+        await __post_to_ledger(url, payload)
+
+        TAA = await __get_taa(controller)
+
+        await __accept_taa(controller, TAA)
+
+        await __assign_pub_did(controller, did_object)
+
```

Comment:
So here we have multiple async calls but they're effectively run synchronously as each one waits for the next one to finish before continuing - but between them there isn't too much shared state...

So could we not call __accept_taa and __assign_pub_did  in tasks and then wait for the tasks to finish... so we won't have to wait for _accpt_taa to finish before running __assign_pub_did?

Having said that there might be some dependency on the back end between them which would make this impossible. I'm just raising the possibility.

So this would become:

```
accept_taa = asyncio.create_task(__accept_taa(controller, TAA))
assign_pub_did = asyncio.create_task(__assign_pub_did(controller, did_object))
await accept_taa
await assign_pub_did
```

So the create_task call will start the method running asynchronously causing each of the __accept_taa and __asign_pub_did to be run asynchrously.

Then awaiting both will ensure that after the second await but have finished before continuing.

Reply:
But of course, this is only possible if these tasks can be run concurrently - might be some reason why they can't.

Reply:
Ah I think I see what you mean. In general, oncurrent will be useful when multiple endpoint calls are made from a multitude of clients so they don't have to wait. But as you correctly assume in this case (kinda per endpoint if you so will) this cannot be concurrent in most cases, because eg here the TAA must have been accepted and the ledger must be aware of that for the controller to be able to write a did. So if we made that cuncurrent we'd get an error if we try writing a did before accepting TAA has finished successfully. But, again for endpoints as a whole, I surely agree, that concurrency will be good

Reply:
Hey looks cool, just having a quick scan. 

assign_pub_did has a dependency on accept_taa. It will fail if the DID on the ledger has not accepted the TAA.